File: day02/main.py
import numpy as np
import pandas as pd

def task1(direction, amount):
    return np.product(amount @ direction)

def task2(direction, amount):
    aim = np.cumsum(amount * direction[:, 1])
    horizontal = amount @ direction[:, 0]
    depth = (aim * amount) @ direction[:, 0]

    return horizontal*depth


if __name__ == '__main__':
    df = pd.read_csv("input.txt", header=None, sep=" ")

    direction = df[0].to_list()
    def dir_to_vec(d):
        return [d == "forward", (d == "down") - (d == "up")]
    direction = np.array(list(map(dir_to_vec, direction)))
    amount = df[1].to_numpy()

    res1 = task1(direction, amount)
    res2 = task2(direction, amount)

    print(f"Task1: {res1}")
    print(f"Task2: {res2}")

File: day02/main.cpp
#include <string>
#include <iostream>
#include <fstream>
#include <array>

int task1(std::ifstream& file) {
    int depth = 0;
    int horizontal = 0;
    std::string direction;
    int amount;

    while (file >> direction) {
        file >> amount;
        if (direction == "forward") {
            horizontal += amount;
        } else if (direction == "down") {
            depth += amount;
        } else if (direction == "up") {
            depth -= amount;
        } else {
            std::cerr << "unrecognized input '" << direction << "'" << std::endl;
        }
    }

    return depth*horizontal;
}

int task2(std::ifstream& file) {
    int aim = 0;
    int depth = 0;
    int horizontal = 0;
    std::string direction;
    int amount;

    while (file >> direction) {
        file >> amount;
        if (direction == "forward") {
            horizontal += amount;
            depth += aim * amount;
        } else if (direction == "down") {
            aim += amount;
        } else if (direction == "up") {
            aim -= amount;
        } else {
            std::cerr << "unrecognized input '" << direction << "'" << std::endl;
        }
    }

    return depth*horizontal;
}
int main() {
    std::ifstream file;
    file.open("input.txt");

    int result1 = task1(file);

    file.clear();
    file.seekg(0, std::ios::beg);

    int result2 = task2(file);

    file.close();

    std::cout << "Result of task1: " << result1 << "\n";
    std::cout << "Result of task2: " << result2 << std::endl;
    return 0;
}
